readthedocs/projects/models.py

```diff
@@ -751,7 +751,8 @@ def versions_from_branch_name(self, branch):
         return (
             self.versions.filter(identifier=branch) |
             self.versions.filter(identifier='remotes/origin/%s' % branch) |
-            self.versions.filter(identifier='origin/%s' % branch)
+            self.versions.filter(identifier='origin/%s' % branch) |
+            self.versions.filter(verbose_name=branch)
```

Comment:
Tags are saved this way: {identifier: hash, verbose_name: tagname}.

GitHub only provides the tag name on the ref.

The only case were I think duplicate values may occur is for branches, since the identifier isn't a hash like they are for tags. I'm going to research a little.

Reply:
These is what I found, looks like the querysets are _really sets_ p:

```
>>> p.versions.filter(identifier=branch)
[<Version: Version awesome of Read The Docs (8)>]
>>> p.versions.filter(verbose_name=branch)
[<Version: Version awesome of Read The Docs (8)>]
>>> p.versions.filter(identifier=branch) | p.versions.filter(verbose_name=branch)
[<Version: Version awesome of Read The Docs (8)>]
```

Reply:
They aren't really sets. In fact, that's why the `distinct` is: https://docs.djangoproject.com/en/2.0/ref/models/querysets/#django.db.models.query.QuerySet.distinct

Anyway, it seems that we are not bitten by this problem here.